Update the build to use emojione 4.5

stof opened the pull request:
Closes #1

Update the build to use emojione 4.5

stof commented:
I added a comparison with the existing JSON file to see the changes. here is the result:

```
Missing version for `wrestlers_tone1`. Was 9.0 previously.
Missing version for `wrestling_tone1`. Was 9.0 previously.
Missing version for `wrestlers_tone2`. Was 9.0 previously.
Missing version for `wrestling_tone2`. Was 9.0 previously.
Missing version for `wrestlers_tone3`. Was 9.0 previously.
Missing version for `wrestling_tone3`. Was 9.0 previously.
Missing version for `wrestlers_tone4`. Was 9.0 previously.
Missing version for `wrestling_tone4`. Was 9.0 previously.
Missing version for `wrestlers_tone5`. Was 9.0 previously.
Missing version for `wrestling_tone5`. Was 9.0 previously.
Missing version for `handshake_tone1`. Was 9.0 previously.
Missing version for `shaking_hands_tone1`. Was 9.0 previously.
Missing version for `handshake_tone2`. Was 9.0 previously.
Missing version for `shaking_hands_tone2`. Was 9.0 previously.
Missing version for `handshake_tone3`. Was 9.0 previously.
Missing version for `shaking_hands_tone3`. Was 9.0 previously.
Missing version for `handshake_tone4`. Was 9.0 previously.
Missing version for `shaking_hands_tone4`. Was 9.0 previously.
Missing version for `handshake_tone5`. Was 9.0 previously.
Missing version for `shaking_hands_tone5`. Was 9.0 previously.
Changed version for `cricket` from 8.0 to 10.0.
Changed version for `flag_black` from 6.0 to 7.0.
Changed version for `waving_black_flag` from 6.0 to 7.0.
Changed version for `flag_white` from 6.0 to 7.0.
Changed version for `waving_white_flag` from 6.0 to 7.0.
Changed version for `rainbow_flag` from 6.0 to 7.0.
Changed version for `gay_pride_flag` from 6.0 to 7.0.
```

We are missing 20 emoji names here, because emojione 3+ removed tone variants for `wrestlers` and `handshake` (the other half being their aliases).
Should I register them as legacy aliases to the base name ? For them, it would keep them at `9.0`

Regarding the changes, the flag changes look legitimate to me. The old build implementation was forcing all flags to be at 6.0, but Unicode added more flags later.

The change for `cricket` is related to https://emojipedia.org/cricket/ (Unicode 10.0) vs https://emojipedia.org/cricket-bat-and-ball/ (Unicode 8.0). Emojione 3.1 reassigned the `:cricket:` shortname to the new emoji in https://github.com/emojione/emojione/commit/4a166f9334809b5044ecf493f7decf47d50e0a8a (which btw made it a BC break in a minor version for anyone storing shortcodes).
I don't know how to handle this to avoid breaking things for gitlab

stof commented:
@MadLittleMods any suggestion about the way to handle this `cricket` case ?

MadLittleMods commented:
> We are missing 20 emoji names here, because emojione 3+ removed tone variants for `wrestlers` and `handshake` (the other half being their aliases).
Should I register them as legacy aliases to the base name ? For them, it would keep them at `9.0`

Yes, let's go with legacy aliases for skin tone variation `wrestlers`, `handshake`, and `shaking_hands`

---

For `cricket`, I guess, we should use the new `10.0` version to keep up to date with the Unicode standard and `emojione`.

On GitLab, if the version isn't supported, it will just fallback to an image which will be fine. When GitLab upgrades `gemojione`, it will have the proper `10.0` version for the new `cricket`

MadLittleMods commented:
Mind sharing your diff script as one of the tasks? Perhaps `compare-emoji-unicode-version-map.js`

MadLittleMods commented on tasks/emoji-unicode-version-map/index.js:
```diff
@@ -37,83 +10,33 @@ const aliasMap = {
-			if(notStringVersionCount > 0) {
-				throw new Error(`unicodeVersion property not a string for ${notStringVersionCount} emojis`);
-			}
-		});
```
I think we should keep this verification step

stof replied:
the way the code is working now does not ever allow having values without a version, as we don't combine multiple sources each providing some versions, but we directly get it from emojione. Do you want to account for cases where the version is missing in emojione itself ?

stof replied:
OK, re-adding the verification step, as there is actually a way to produce a violation: having an alias pointing to an non-existent name.

MadLittleMods commented on tasks/emoji-unicode-version-map/index.js:
```diff
@@ -37,83 +10,33 @@ const aliasMap = {
-				throw new Error(`unicodeVersion property not a string for ${notStringVersionCount} emojis`);
-			}
-		});
+		const unicodeVersion = emoji.unicode_version.toFixed(1);
```
What is `toFixed(1)` accomplishing here?

Perhaps there will be Unicode standards with smaller decimals in the future 🤷‍♂️

stof replied:
converting a number to a string with 1 decimal

stof replied:
what I could do is using `Intl.NumberFormat` to format the number with *at least* 1 decimal instead, but that would add such a dependency for an hypothetical future Unicode change.

Another solution would be to add a validation that we don't loose precision here, to make the script fail the day Unicode makes such versioning change (and not add extra dependencies until that time).
From what I can tell, it seems like Unicode even stopped using the first decimal (all recent Unicode versions are in `.0`. According to the list in the https://emojipedia.org/ footer, the last version using non-zero decimal was `6.1`.

stof replied:
Implemented as a verification check, so that we know about the issue when running the build if the Unicode standards ever use multiple decimals in their versions.

MadLittleMods commented on tasks/emoji-unicode-version-map/index.js:
```diff
@@ -37,83 +10,33 @@ const aliasMap = {
-			});
-		});
+		const emojiShortName = emoji.shortname;
+		const emojiNameKey = emojiShortName.slice(1, emojiShortName.length - 1);
```
We can just use the `emoji.name` AFAICT

stof replied:
not if you want it to be compatible with emojione short names (`name`  is the Unicode name AFAICT)

stof replied:
confirmed. For instance, one of them has the name `palms up together: dark skin tone` while its emojione short name is `:palms_up_together_tone5:`

MadLittleMods replied:
Thanks for looking into it!

stof commented:
> Mind sharing your diff script as one of the tasks? Perhaps `compare-emoji-unicode-version-map.js`

That would require keeping the old JSON map in the repo though, a my script compares the 2 JSON files. Or maybe we could make that an argument of the script to let you provide it from another source ?

stof commented:
> > Mind sharing your diff script as one of the tasks? Perhaps `compare-emoji-unicode-version-map.js`
> 
> That would require keeping the old JSON map in the repo though, a my script compares the 2 JSON files. Or maybe we could make that an argument of the script to let you provide it from another source ?

This is now implemented as a script expecting the path to the old JSON map as an argument (see the commit message introducing it) as keeping the old JSON map committed does not make sense IMO.

stof commented:
This is now ready. Output of the comparison script:

```
Changed version for `cricket` from 8.0 to 10.0.
Changed version for `flag_black` from 6.0 to 7.0.
Changed version for `waving_black_flag` from 6.0 to 7.0.
Changed version for `flag_white` from 6.0 to 7.0.
Changed version for `waving_white_flag` from 6.0 to 7.0.
Changed version for `rainbow_flag` from 6.0 to 7.0.
Changed version for `gay_pride_flag` from 6.0 to 7.0.

Summary:
- missing 0 emojis
- version changes for 7 emojis
```

stof commented:
> In GitLab, if the version isn't supported, it will just fallback to an image which will be fine. When GitLab upgrades `gemojione`, it will have the proper `10.0` version for the new `cricket`

The main issue for Gitlab in this update would be that `:cricket:` changed meaning, in case Gitlab stores emojis as their shortcodes rather than native Unicode emojis.
but in the meantime, having one emoji being classified at a higher version will indeed not break things (the opposite might)

MadLittleMods commented on tasks/emoji-unicode-version-map/index.js:
```diff
@@ -1,119 +1,88 @@
+		const unicodeVersion = emoji.unicode_version.toFixed(1);
+
+		if (emoji.unicode_version !== parseFloat(unicodeVersion)) {
+			throw new Error(`Unsupported Unicode version ${emoji.unicode_version}`);
```
I feel like we should add a comment to clarify this better.

```
throw new Error(`Unsupported Unicode version "${emoji.unicode_version}" parsed as ${unicodeVersion}. We probably need to update the number to string conversion just above`);
```

MadLittleMods commented on tasks/emoji-unicode-version-map/index.js:
```diff
@@ -1,119 +1,88 @@
-			});
-		});
-		return emojiToUnicodeVersionMap;
+		if(version === undefined) {
```
Instead of checking for `undefined`, probably a bit safer to just go with `if(!version)`. We can also do this in `tasks/compare-maps.js`

MadLittleMods replied:
Since we had this before, let's just go with it 🚀

MadLittleMods commented on tasks/compare-maps.js:
```diff
@@ -0,0 +1,31 @@
+const fs = require('fs');
+
+const argv = require('yargs/yargs')(process.argv.slice(2)).argv;
```
AFAICT, we can simplify to the following

```js
const yargs = require('yargs');
yargs.argv._[0]
```

MadLittleMods commented:
Thanks for the updates @stof! I only spotted a few more minor nits

MadLittleMods commented:
I am going to work on add the last bits and merge this in 💪

MadLittleMods commented:
Thanks @stof, this is now shipped in v0.3.0 🚀 , https://www.npmjs.com/package/emoji-unicode-version
